sdk/attestation/azure-security-attestation/src/attestation_administration_client.cpp

```diff
@@ -126,6 +122,163 @@ AttestationAdministrationClient::GetAttestationPolicy(
   return Response<AttestationToken<std::string>>(returnedToken, std::move(response));
 }
 
+Models::AttestationToken<std::nullptr_t>
+AttestationAdministrationClient::CreateSetAttestationPolicyToken(
```

Comment:
is this API, not doing any I/O ?  or I missed it?

Reply:
This API does not do any I/O. It is a helper function which allows customers to create an AttestationToken<> which can be used to calculate the hash of the JWS which was sent to the service.

Reply:
One of those `magic API methods` .  I am fine with them, but be ready to hear questions and controversy around `magic API methods`.  As it was mentioned on last stand up meeting, C++ tries to keep a correlation between each SDK client method to a REST call.  And then let any other <required/extra step/validation> to customers.

Reply:
I can add a client class derived from `AttestationToken<nullptr_t>` if you would prefer - that's similar to what has been done in other languages. It's a much clunkier experience and is completely undiscoverable (this feedback came from Java's review). But there's never going to be a wire experience around this class. And it is *not* a model type, it's a service type.

Reply:
One other question: Do you have a working definition for `magic API method`? Are there guidelines which help define them and explain why they should be avoided? Why isn't the `ClientVersion` API also considered a `magic API method`?

And my interpretation of Jeff's comment is that SDK methods which interact with the service should execute one and only one HTTP operation. But that says nothing about helper methods and other constructs which don't interact with the service.

My taxonomy of methods is:
* C++ Infrastructure (constructors and destructors) - these should never interact with the service directly.
* Service methods - these are methods which interact with the service and should result in at most one HTTP call
* Client methods - these are methods which do NOT interact with the service. These can exist if they solve a problem (however I have no idea why `ClientVersion` exists, to be honest - for the life of me, I don't know why a customer would need to know the version they chose when they configured the client).

Reply:
> Do you have a working definition for magic API method?

I don't. At least not one officially approved.  I can tell about what have been considered `magic` in the past. Like:

- Key Vault local crypto client.  While other languages provide support for running crypto operations locally (crypto client tries to run locally first, and if it can't, it sends the request to the service), when I was implementing this for C++, it was considered `magic` and I removed the local support and made the crypto client to always send the request.

- Key Vault `DownloadCertificate` API for Key Vault Certificate.  This was considered a `magic` API because it requires 2 API calls (GetCertificate + GetSecret) and it was moved out of the public API to be just shown as part of the test cases (so people know how they can implement it).  The main `const` for this API was that Response<T> returns only one raw-response and the API is generating two raw-responses. All other languages provide the magic.

- Key Vault LRO `CreateCertificate`. This is still under debate. While other languages makes the LRO to get the Certificate when the LRO is completed, C++ consider that as magic and it's trying to go against it.  Instead, the customer should call `GetCertificate` once the LRO.isCompleted() is true.


Coming back to this specific case. When a client operation is not performing I/O or at least returning some internal state form the client (like the service version or url); it should not be part of the client.  I would move that method to be part of some Util model like `Attestation::<Namespace>::<UtilsClassName>::Foo(..., ... , .. );`  .  But I don't usually make the rules for this.  You want to check this with @JeffreyRichter

Reply:
That helps a bunch. At some point, I asked about free functions in the SDK (that's how we solve this problem for JS - the `CreateSetPolicyToken` API is a function defined on the global object , I got the impression that we weren't encouraging flat functions. I *don't* want to create an entire client class for a single function - I'd MUCH prefer a simple function to implement this.